File: supervisorpt/src/specificworker.py
#
# Copyright (C) 2016 by YOUR NAME HERE
#
#    This file is part of RoboComp
#
#    RoboComp is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    RoboComp is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
#

import sys, os, Ice, traceback, time

import networkx as nx
import matplotlib.pyplot as plt

from PySide import *
from genericworker import *

ROBOCOMP = ''
try:
	ROBOCOMP = os.environ['ROBOCOMP']
except:
	print '$ROBOCOMP environment variable not set, using the default value /opt/robocomp'
	ROBOCOMP = '/opt/robocomp'
if len(ROBOCOMP)<1:
	print 'genericworker.py: ROBOCOMP environment variable not set! Exiting.'
	sys.exit()


preStr = "-I"+ROBOCOMP+"/interfaces/ --all "+ROBOCOMP+"/interfaces/"
Ice.loadSlice(preStr+"DifferentialRobot.ice")
from RoboCompDifferentialRobot import *
Ice.loadSlice(preStr+"GotoPoint.ice")
from RoboCompGotoPoint import *

class SpecificWorker(GenericWorker):
  
	posiciones={}
	g=nx.Graph()
	estado='INIT'
	nodos=[]
	ruta=[67]
	print "Ruta: ", ruta
	
	def __init__(self, proxy_map):
		super(SpecificWorker, self).__init__(proxy_map)
		self.timer.timeout.connect(self.compute)
		self.Period = 2000
		self.timer.start(self.Period)
		self.grafo()
		self.compute()
		
	def setParams(self, params):
		return True
	
	def init(self, lista):
	 self.estado="TI"
	  
	  
	def ti(self, lista):
	  if len(self.ruta)==0:
	      self.estado="INIT"
	      return
	  
	  print "To TI"
	  
	  self.nodos=nx.shortest_path(self.g,source=str(self.nodoCercano()),target=str(self.ruta[0]))
	  self.ruta.pop(0)
	  print self.nodos
	  
	  self.estado="PI"
	  
	def pi(self, lista):
	  print "To PI"	  
	  if len(self.nodos)==0:
	      self.estado="INIT"
	      return
	    
	  primernodo=self.nodos[0]
	  self.nodos.pop(0)
	  
	  print "Pos robot ", self.posiciones[primernodo][0], self.posiciones[primernodo][1]
	  self.gotopoint_proxy.go("",self.posiciones[primernodo][0],self.posiciones[primernodo][1],0.3)
	  
	  self.estado="GO"
	  
	  
	def go(self, lista):
	  print "To GO"
	  
	  if self.gotopoint_proxy.atTarget():
		print "Target"
		self.estado="PI"
		return
	  
	  else:
		print "Searching"
	  	 
	def grafo(self):   
	  with open("puntos.txt","r") as f:
	    for line in f:
	      l=line.strip("\n").split()
	      if l[0]=="N":
		self.g.add_node(l[1], x= float(l[2]),y=float(l[3]),name="")
		self.posiciones[l[1]] = (float(l[2]),float(l[3]))
	      else:
		self.g.add_edge(l[1], l[2])
		  
          #print self.posiciones
          img = plt.imread("plano.png")
          plt.imshow(img, extent = [-12284,25600,-3840,9023])
          '''nx.draw_networkx(self.g, self.posiciones)
          
          print "Haciendo camino minimo"
          print nx.shortest_path(self.g,source="1", target="6")
          plt.show()'''
	  
	@QtCore.Slot()
	
	def compute(self):
	  bState=TBaseState()
	  
	  state = {
	    'INIT': {"command": self.init, "params": []},
	    'TI': {"command": self.ti, "params": []},
	    'PI': {"command": self.pi, "params": []},
	    'GO': {"command": self.go, "params": []}}
	  state[self.estado]["command"](state[self.estado]["params"])

	def nodoCercano(self):
	  bState = TBaseState()
	  bState = self.differentialrobot_proxy.getBaseState()
	  r = (bState.x , bState.z)
	  dist = lambda r,n: (r[0]-n[0])**2+(r[1]-n[1])**2
	  #funcion que devuele el nodo mas cercano al robot
	  print "Nodo/s mas cercano/s"
	  return  sorted(list (( n[0] ,dist(n[1],r)) for n in self.posiciones.items() ), key=lambda s: s[1])[0][0]  

File: supervisor/src/specificworker.cpp
/*
 *    Copyright (C) 2016 by YOUR NAME HERE
 *
 *    This file is part of RoboComp
 *
 *    RoboComp is free software: you can redistribute it and/or modify
 *    it under the terms of the GNU General Public License as published by
 *    the Free Software Foundation, either version 3 of the License, or
 *    (at your option) any later version.
 *
 *    RoboComp is distributed in the hope that it will be useful,
 *    but WITHOUT ANY WARRANTY; without even the implied warranty of
 *    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *    GNU General Public License for more details.
 *
 *    You should have received a copy of the GNU General Public License
 *    along with RoboComp.  If not, see <http://www.gnu.org/licenses/>.
 */
#include "specificworker.h"

/**
* \brief Default constructor
*/
SpecificWorker::SpecificWorker ( MapPrx& mprx ) : GenericWorker ( mprx )
{

}

/**
* \brief Default destructor
*/
SpecificWorker::~SpecificWorker()
{

}

bool SpecificWorker::setParams ( RoboCompCommonBehavior::ParameterList params )
{


        iner = new InnerModel ( "/home/jesusuiano/robocomp/files/innermodel/simpleworld.xml" );
        //timer.start ( Period );
        current = 0;

        return true;
}

void SpecificWorker::compute()
{

        try {
                RoboCompDifferentialRobot::TBaseState bState;
                differentialrobot_proxy->getBaseState ( bState );
                qDebug() << bState.x << bState.z;
                iner->updateTransformValues ( "base",bState.x,0,bState.z,0,bState.alpha,0 );
                qDebug() << current << "current state";
                switch ( estado ) {
                case Estado::BUSCAR:
                        qDebug() << "State SEARCH" ;
                        if ( tag.getID() == current ) {
                                gotopoint_proxy->stop();
                                gotopoint_proxy->go ( "base", tag.getPose().x(), tag.getPose().z(),0 );
                                estado = Estado::ESPERAR; 
                                qDebug() <<"State change to WAIT";
                        } else
                                gotopoint_proxy->turn ( 0.6 );
                        break;
                case Estado::ESPERAR:
                        qDebug() << "State WAIT" ;
                        if ( gotopoint_proxy->atTarget() == true ) {
                                gotopoint_proxy->stop();
                                estado = Estado::BUSCAR;
                                qDebug() <<"State change to SEARCH";
                                current = ( current+1 ) %4;
                        } else if ( tag.cambiado() ) {
                                gotopoint_proxy->go ( "", tag.getPose().x(), tag.getPose().z(), 0 );
                        }
                        break;
                }

        } catch ( const Ice::Exception &e ) {
                std::cout << "Error reading from Camera" << e << std::endl;
        }
}



void SpecificWorker::newAprilTag ( const tagsList &tags )
{
        tag.copy ( tags[0].tx,tags[0].tz,tags[0].id );
}






